### ISSUE TITLE
Java application not showing up in attach menu

### ISSUE THREAD

**Issue by xrnss** [Col-E/Recaf#399]
I'm pretty sure it should, as when I write my own Java code to show all available VMs, it shows up. I'm not sure what else I can say but is there any reason it might not be showing up? The application name is Lunar Client (minecraft thing if anyone knows). However, I can see Forge 1.8.9 but not Lunar

**Comment by xrnss** [comment]
Maybe add a text input for PID, and then it can search for the application with that PID that way?

**Comment by win32kbase** [comment]
This isn't an issue with Recaf, the client is launched with the DisableAttachMechanism argument. You will have to launch the client without this launch argument for Recaf to be able to attach to it.
![image](https://user-images.githubusercontent.com/6201484/129490563-7f46c83f-3870-4df5-ba3f-3c09816c08a8.png)

Removing this flag myself allows me to attach with Recaf:
![image](https://user-images.githubusercontent.com/6201484/129490891-7beb137f-abae-408f-9fe8-c14ed0fa0cc3.png)

**Comment by xrnss** [comment]
tysfm i'll test it tmr and let you know. i knew the app had this tag but i thought i would have to "hack" the electron app to remove it (the electron app also has these tags however if you can get it to work by modifying the exe file then maybe they do nothing). again thank you for the quick response and solution.

**Comment by win32kbase** [comment]
> tysfm i'll test it tmr and let you know. i knew the app had this tag but i thought i would have to "hack" the electron app to remove it (the electron app also has these tags however if you can get it to work by modifying the exe file then maybe they do nothing). again thank you for the quick response and solution.

I can record a video showing how I edited the launcher if that would help you

**Comment by xrnss** [comment]
Did you edit the launcher code (the electron one) or just the properties menu of the exe file (i know the location i've been snooping around this stuff for ages LOOL)

**Comment by xrnss** [comment]
> > tysfm i'll test it tmr and let you know. i knew the app had this tag but i thought i would have to "hack" the electron app to remove it (the electron app also has these tags however if you can get it to work by modifying the exe file then maybe they do nothing). again thank you for the quick response and solution.
> 
> I can record a video showing how I edited the launcher if that would help you

oh i didn't realise quote reply stop as a thing

**Comment by xrnss** [comment]
> > tysfm i'll test it tmr and let you know. i knew the app had this tag but i thought i would have to "hack" the electron app to remove it (the electron app also has these tags however if you can get it to work by modifying the exe file then maybe they do nothing). again thank you for the quick response and solution.
> 
> I can record a video showing how I edited the launcher if that would help you

honestly yes would love that if you have free time if not it's completely fine

**Comment by win32kbase** [comment]
> > > tysfm i'll test it tmr and let you know. i knew the app had this tag but i thought i would have to "hack" the electron app to remove it (the electron app also has these tags however if you can get it to work by modifying the exe file then maybe they do nothing). again thank you for the quick response and solution.
> > 
> > 
> > I can record a video showing how I edited the launcher if that would help you
> 
> honestly yes would love that if you have free time if not it's completely fine

https://youtu.be/XVz7buhPj6U

**Comment by xrnss** [comment]
thank you so much, i didn't realise that was in process hacker LOL. yes i've unpacked that before and removed those strings but for whatever reason i never thought of repacking it. thank you so much for the help dude

**Comment by win32kbase** [comment]
> thank you so much, i didn't realise that was in process hacker LOL. yes i've unpacked that before and removed those strings but for whatever reason i never thought of repacking it. thank you so much for the help dude

no problem

**Comment by Col-E** [comment]
Looks like this is resolved then

**Comment by xrnss** [comment]
After fixing the attach disabled and trying to attach, I ran into the following error: com.sun.tools.attach.AgentInitializationException: Agent JAR loaded but agent failed to initialize

I then went to the logs of the game and found this. Could any help be provided as I am quite confused?

```
[2021-08-16 08:55:35.545] [info]  [OUTPUT] 08:55:35.541 [Attach Listener] WARN : Failed to delete old log file, will append instead

[2021-08-16 08:55:35.548] [info]  [OUTPUT] 08:55:35.545 [Attach Listener] INFO : Recaf-2.20.1
08:55:35.546 [Attach Listener] INFO : Starting as agent...

[2021-08-16 08:55:35.700] [info]  [OUTPUT] 08:55:35.699 [Attach Listener] TRACE: Loading configuration

[2021-08-16 08:55:36.104] [info]  [OUTPUT] 08:55:36.103 [Attach Listener] INFO : Loaded instrumentation workspace

[2021-08-16 08:55:36.107] [info]  [OUTPUT] 08:55:36.106 [Attach Listener] INFO : Loaded 1 plugins

[2021-08-16 08:55:36.111] [info]  [OUTPUT] Exception in thread "Attach Listener" java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)

[2021-08-16 08:55:36.112] [info]  [OUTPUT] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndStartAgent(Unknown Source)
	at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndCallAgentmain(Unknown Source)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at me.coley.recaf.Agent.agent(Agent.java:53)
	at me.coley.recaf.Agent.agentmain(Agent.java:43)
	... 6 more
Caused by: java.lang.RuntimeException: Unable to initialize toolkit
	at me.coley.recaf.util.VMUtil.tkIint(VMUtil.java:184)
	at me.coley.recaf.ui.MainWindow.get(MainWindow.java:198)
	at me.coley.recaf.control.gui.GuiController.run(GuiController.java:44)
	at me.coley.recaf.command.impl.Initializer.startController(Initializer.java:73)
	at me.coley.recaf.Recaf.launch(Recaf.java:112)
	at me.coley.recaf.Recaf.agent(Recaf.java:72)
	... 12 more
Caused by: java.lang.IllegalStateException: Toolkit already initialized
	at javafx.graphics/com.sun.javafx.application.PlatformImpl.startup(Unknown Source)
	at javafx.graphics/javafx.application.Platform.startup(Unknown Source)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at me.coley.recaf.util.VMUtil.tkIint(VMUtil.java:180)
	... 17 more
Agent failed to start!
```

**Comment by xrnss** [comment]
After digging around, does this mean JavaFX was already initialized? If so, how do I stop Recaf trying to load it then???

**Comment by xrnss** [comment]
I believe this was caused because I pressed attach twice. I am going to see if there is a different error.

**Comment by xrnss** [comment]
Okay, I've found the real error here. How would I go about fixing this?

`[2021-08-16 09:15:56.064] [info]  [OUTPUT] 09:15:56.059 [Attach Listener] WARN : Failed to delete old log file, will append instead
09:15:56.062 [Attach Listener] INFO : Recaf-2.20.1
09:15:56.063 [Attach Listener] INFO : Starting as agent...

[2021-08-16 09:15:56.296] [info]  [OUTPUT] 09:15:56.295 [Attach Listener] TRACE: Loading configuration

[2021-08-16 09:15:59.481] [info]  [OUTPUT] 09:15:59.480 [Attach Listener] INFO : Loaded instrumentation workspace

[2021-08-16 09:15:59.496] [info]  [OUTPUT] 09:15:59.489 [Attach Listener] INFO : Loaded 1 plugins

[2021-08-16 09:16:01.873] [info]  [OUTPUT] Exception in thread "JavaFX Application Thread" java.lang.RuntimeException: java.lang.NullPointerException: Cannot invoke "java.lang.ClassLoader.getResource(String)" because "<local2>" is null

[2021-08-16 09:16:01.884] [info]  [OUTPUT] 	at me.coley.recaf.ui.MainWindow.lambda$get$2(MainWindow.java:205)
	at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runLater$10(Unknown Source)
	at java.base/java.security.AccessController.doPrivileged(Unknown Source)
	at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runLater$11(Unknown Source)
	at javafx.graphics/com.sun.glass.ui.InvokeLaterDispatcher$Future.run(Unknown Source)
	at javafx.graphics/com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
	at javafx.graphics/com.sun.glass.ui.win.WinApplication.lambda$runLoop$3(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.lang.NullPointerException: Cannot invoke "java.lang.ClassLoader.getResource(String)" because "<local2>" is null
	at javafx.graphics/com.sun.javafx.css.StyleManager.getURL(Unknown Source)
	at javafx.graphics/com.sun.javafx.css.StyleManager.loadStylesheetUnPrivileged(Unknown Source)
	at javafx.graphics/com.sun.javafx.css.StyleManager.loadStylesheet(Unknown Source)
	at javafx.graphics/com.sun.javafx.css.StyleManager.processStylesheets(Unknown Source)
	at javafx.graphics/com.sun.javafx.css.StyleManager.gatherSceneStylesheets(Unknown Source)
	at javafx.graphics/com.sun.javafx.css.StyleManager.findMatchingStyles(Unknown Source)
	at javafx.graphics/javafx.scene.CssStyleHelper.createStyleHelper(Unknown Source)
	at javafx.graphics/javafx.scene.Node.reapplyCss(Unknown Source)
	at javafx.graphics/javafx.scene.Node.doProcessCSS(Unknown Source)
	at javafx.graphics/javafx.scene.Node$1.doProcessCSS(Unknown Source)
	at javafx.graphics/com.sun.javafx.scene.NodeHelper.processCSSImpl(Unknown Source)
	at javafx.graphics/com.sun.javafx.scene.ParentHelper.superProcessCSSImpl(Unknown Source)
	at javafx.graphics/com.sun.javafx.scene.ParentHelper.superProcessCSS(Unknown Source)
	at javafx.graphics/javafx.scene.Parent.doProcessCSS(Unknown Source)
	at javafx.graphics/javafx.scene.Parent$1.doProcessCSS(Unknown Source)
	at javafx.graphics/com.sun.javafx.scene.ParentHelper.processCSSImpl(Unknown Source)
	at javafx.graphics/com.sun.javafx.scene.NodeHelper.processCSS(Unknown Source)
	at javafx.graphics/javafx.scene.Node.processCSS(Unknown Source)
	at javafx.graphics/javafx.scene.Scene.doCSSPass(Unknown Source)
	at javafx.graphics/javafx.scene.Scene.preferredSize(Unknown Source)
	at javafx.graphics/javafx.scene.Scene$2.preferredSize(Unknown Source)
	at javafx.graphics/com.sun.javafx.scene.SceneHelper.preferredSize(Unknown Source)
	at javafx.graphics/javafx.stage.Window$12.invalidated(Unknown Source)
	at javafx.base/javafx.beans.property.BooleanPropertyBase.markInvalid(Unknown Source)
	at javafx.base/javafx.beans.property.BooleanPropertyBase.set(Unknown Source)
	at javafx.graphics/javafx.stage.Window.setShowing(Unknown Source)
	at javafx.graphics/javafx.stage.Window.show(Unknown Source)
	at javafx.graphics/javafx.stage.Stage.show(Unknown Source)
	at me.coley.recaf.ui.MainWindow.start(MainWindow.java:63)
	at me.coley.recaf.ui.MainWindow.lambda$get$2(MainWindow.java:203)
	... 7 more

[2021-08-16 09:16:02.251] [info]  [OUTPUT] Exception in thread "JavaFX Application Thread" 
[2021-08-16 09:16:02.260] [info]  [OUTPUT] java.lang.NullPointerException: Cannot invoke "java.lang.ClassLoader.getResource(String)" because "<local2>" is null
	at javafx.graphics/com.sun.javafx.css.StyleManager.getURL(Unknown Source)
	at javafx.graphics/com.sun.javafx.css.StyleManager.loadStylesheetUnPrivileged(Unknown Source)
	at javafx.graphics/com.sun.javafx.css.StyleManager.loadStylesheet(Unknown Source)
	at javafx.graphics/com.sun.javafx.css.StyleManager.processStylesheets(Unknown Source)
	at javafx.graphics/com.sun.javafx.css.StyleManager.gatherSceneStylesheets(Unknown Source)
	at javafx.graphics/com.sun.javafx.css.StyleManager.findMatchingStyles(Unknown Source)
	at javafx.graphics/javafx.scene.CssStyleHelper.createStyleHelper(Unknown Source)
	at javafx.graphics/javafx.scene.Node.reapplyCss(Unknown Source)
	at javafx.graphics/javafx.scene.Node.doProcessCSS(Unknown Source)
	at javafx.graphics/javafx.scene.Node$1.doProcessCSS(Unknown Source)
	at javafx.graphics/com.sun.javafx.scene.NodeHelper.processCSSImpl(Unknown Source)
	at javafx.graphics/com.sun.javafx.scene.ParentHelper.superProcessCSSImpl(Unknown Source)
	at javafx.graphics/com.sun.javafx.scene.ParentHelper.superProcessCSS(Unknown Source)
	at javafx.graphics/javafx.scene.Parent.doProcessCSS(Unknown Source)
	at javafx.graphics/javafx.scene.Parent$1.doProcessCSS(Unknown Source)
	at javafx.graphics/com.sun.javafx.scene.ParentHelper.processCSSImpl(Unknown Source)
	at javafx.graphics/com.sun.javafx.scene.NodeHelper.processCSS(Unknown Source)
	at javafx.graphics/javafx.scene.Node.processCSS(Unknown Source)
	at javafx.graphics/javafx.scene.Scene.doCSSPass(Unknown Source)
	at javafx.graphics/javafx.scene.Scene$ScenePulseListener.pulse(Unknown Source)
	at javafx.graphics/com.sun.javafx.tk.Toolkit.lambda$runPulse$2(Unknown Source)
	at java.base/java.security.AccessController.doPrivileged(Unknown Source)
	at javafx.graphics/com.sun.javafx.tk.Toolkit.runPulse(Unknown Source)
	at javafx.graphics/com.sun.javafx.tk.Toolkit.firePulse(Unknown Source)
	at javafx.graphics/com.sun.javafx.tk.quantum.QuantumToolkit.pulse(Unknown Source)
	at javafx.graphics/com.sun.javafx.tk.quantum.QuantumToolkit.pulse(Unknown Source)
	at javafx.graphics/com.sun.javafx.tk.quantum.QuantumToolkit.pulseFromQueue(Unknown Source)
	at javafx.graphics/com.sun.javafx.tk.quantum.QuantumToolkit.lambda$runToolkit$11(Unknown Source)
	at javafx.graphics/com.sun.glass.ui.InvokeLaterDispatcher$Future.run(Unknown Source)
	at javafx.graphics/com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
	at javafx.graphics/com.sun.glass.ui.win.WinApplication.lambda$runLoop$3(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)`

**Comment by xrnss** [comment]
This error also causes the application to crash

**Comment by xrnss** [comment]
I've tested this using Java 16 and Java 8, both cause the same error. Seems to be some weird issue with JavaFX

**Comment by Col-E** [comment]
Looks like a different error. Can open a new issue for this one.

**Comment by xrnss** [comment]
I will

**Comment by smallshen** [comment]
@xrnss lunar client will set "DisableAttachMechanism" in the envirament variables, you need to make sure you removed that as well

**Comment by xrnss** [comment]
That is not anything to do with the error.. how do you think I even got it to try attach without removing that? Yes I have removed it but that does not relate to the erorr

**Comment by smallshen** [comment]
are you trying to get lunar jar classes?

**Comment by xrnss** [comment]
This issue is resolved, see #400 for the current issue
